File: services/complaint-service/app/sentiment.py
"""Lightweight keyword sentiment + urgency (no ML cloud)."""
import re

CRITICAL = [
    "urgent", "emergency", "danger", "dangerous", "toxic", "flood", "overflow",
    "blocked drain", "sewage", "disease", "children", "hospital", "critical",
    "immediately", "hazard", "fire", "dead animal",
]
HIGH = [
    "smelly", "stinking", "garbage pile", "not cleaned", "days", "week",
    "angry", "frustrated", "worst", "horrible", "disgusting", "rats", "mosquito",
]
POSITIVE = ["thank", "resolved", "clean", "good", "appreciate", "happy", "improved"]


def analyze_text(text: str) -> dict:
    t = (text or "").lower()
    score = 0
    for w in POSITIVE:
        if w in t:
            score += 1
    for w in HIGH:
        if w in t:
            score -= 1
    for w in CRITICAL:
        if w in t:
            score -= 2

    if score <= -3:
        label, urgency = "very_negative", "critical"
    elif score <= -1:
        label, urgency = "negative", "high"
    elif score >= 2:
        label, urgency = "positive", "low"
    else:
        label, urgency = "neutral", "medium"

    # Exclamation / caps intensity
    if text.count("!") >= 2 or re.search(r"[A-Z]{6,}", text or ""):
        if urgency == "medium":
            urgency = "high"
        elif urgency == "high":
            urgency = "critical"
        score -= 1
        if label == "neutral":
            label = "negative"

    return {
        "sentiment_score": float(score),
        "sentiment_label": label,
        "urgency": urgency,
    }

File: frontend/src/lib/sentiment.ts
import { useMemo } from "react";

const CRITICAL = [
  "urgent", "emergency", "danger", "toxic", "flood", "sewage", "disease",
  "critical", "immediately", "hazard", "dead animal",
];
const HIGH = [
  "smelly", "stinking", "garbage", "angry", "frustrated", "worst",
  "horrible", "disgusting", "rats", "mosquito", "not cleaned",
];
const POSITIVE = ["thank", "resolved", "clean", "good", "appreciate", "happy"];

export type SentimentResult = {
  score: number;
  label: string;
  urgency: "low" | "medium" | "high" | "critical";
};

export function analyzeComplaintText(text: string): SentimentResult {
  const t = (text || "").toLowerCase();
  let score = 0;
  for (const w of POSITIVE) if (t.includes(w)) score += 1;
  for (const w of HIGH) if (t.includes(w)) score -= 1;
  for (const w of CRITICAL) if (t.includes(w)) score -= 2;

  let label = "neutral";
  let urgency: SentimentResult["urgency"] = "medium";
  if (score <= -3) {
    label = "very_negative";
    urgency = "critical";
  } else if (score <= -1) {
    label = "negative";
    urgency = "high";
  } else if (score >= 2) {
    label = "positive";
    urgency = "low";
  }

  if ((text.match(/!/g) || []).length >= 2) {
    if (urgency === "medium") urgency = "high";
    else if (urgency === "high") urgency = "critical";
  }

  return { score, label, urgency };
}

export function useLiveSentiment(text: string): SentimentResult {
  return useMemo(() => analyzeComplaintText(text), [text]);
}

export function urgencyColor(u: string) {
  switch (u) {
    case "critical":
      return "bg-red-600 text-white";
    case "high":
      return "bg-orange-500 text-white";
    case "low":
      return "bg-emerald-100 text-emerald-800";
    default:
      return "bg-amber-100 text-amber-900";
  }
}
